hydra-node/test/Hydra/NetworkSpec.hs

```diff
@@ -34,61 +34,78 @@ import Test.QuickCheck (
+        [port1, port2, port3] <- fmap fromIntegral <$> randomUnusedTCPPorts 3
+        withOuroborosNetwork tracer (lo, port1) [(lo, port3), (lo, port2)] (atomically . writeTQueue node1received) $ \hn1 ->
+          withOuroborosNetwork tracer (lo, port2) [(lo, port1), (lo, port3)] (atomically . writeTQueue node2received) $ \hn2 -> do
+            withOuroborosNetwork tracer (lo, port3) [(lo, port1), (lo, port2)] (atomically . writeTQueue node3received) $ \hn3 -> do
+              assertAllNodesBroadcast port1 port2 port3 hn1 hn2 hn3 node1received node2received node3received
 
   describe "0MQ Network" $
     it "broadcasts messages between 3 connected peers" $ do
-      node1received <- newEmptyTMVarIO
-      node2received <- newEmptyTMVarIO
-      node3received <- newEmptyTMVarIO
+      node1received <- atomically newTQueue
+      node2received <- atomically newTQueue
+      node3received <- atomically newTQueue
       showLogsOnFailure $ \tracer -> failAfter 10 $ do
-        withZeroMQNetwork tracer (lo, 55677) [(lo, 55678), (lo, 55679)] (atomically . putTMVar node1received) $ \hn1 ->
-          withZeroMQNetwork tracer (lo, 55678) [(lo, 55677), (lo, 55679)] (atomically . putTMVar node2received) $ \hn2 ->
-            withZeroMQNetwork tracer (lo, 55679) [(lo, 55677), (lo, 55678)] (atomically . putTMVar node3received) $ \hn3 -> do
-              concurrently_ (assertBroadcastFrom requestTx hn1 [node2received, node3received]) $
-                concurrently_
-                  (assertBroadcastFrom requestTx hn2 [node1received, node3received])
-                  (assertBroadcastFrom requestTx hn3 [node2received, node1received])
+        [port1, port2, port3] <- fmap fromIntegral <$> randomUnusedTCPPorts 3
+        withZeroMQNetwork tracer (lo, port1) [(lo, port2), (lo, port3)] (atomically . writeTQueue node1received) $ \hn1 ->
+          withZeroMQNetwork tracer (lo, port2) [(lo, port1), (lo, port3)] (atomically . writeTQueue node2received) $ \hn2 ->
+            withZeroMQNetwork tracer (lo, port3) [(lo, port1), (lo, port2)] (atomically . writeTQueue node3received) $ \hn3 -> do
+              assertAllNodesBroadcast port1 port2 port3 hn1 hn2 hn3 node1received node2received node3received
 
   describe "Serialisation" $ do
     it "can roundtrip CBOR encoding/decoding of HydraMessage" $ property $ prop_canRoundtripCBOREncoding @(HydraMessage SimpleTx)
 
-assertBroadcastFrom :: (Eq a, Show a) => a -> Network IO a -> [TMVar IO a] -> IO ()
-assertBroadcastFrom requestTx network receivers =
-  tryBroadcast `catch` \(_ :: HUnitFailure) -> tryBroadcast
- where
-  tryBroadcast = do
-    broadcast network requestTx
-    forM_ receivers $ \var -> failAfter 1 $ atomically (takeTMVar var) `shouldReturn` requestTx
+assertAllNodesBroadcast ::
+  PortNumber ->
+  PortNumber ->
+  PortNumber ->
+  Network IO Integer ->
+  Network IO Integer ->
+  Network IO Integer ->
+  TQueue IO Integer ->
+  TQueue IO Integer ->
+  TQueue IO Integer ->
```

Comment:
:sweat_smile: ... nice signature... I guess we could generalize it to `n` nodes although it's not strictly needed and work as is.

Reply:
Yes, it's not great but doing the generic thing will be more work I think 🤔 I will give it a try though.